src/memory/buffer_pool.cpp

```diff
@@ -6,30 +6,95 @@ SMILE_NS_BEGIN
 
 BufferPool::BufferPool( FileStorage* storage, const BufferPoolConfig& config ) noexcept {
 	p_storage = storage;
-	p_pool = (char*) malloc( config.m_poolSizeKB*1024*sizeof(char) );
+	m_config = config;	
+}
+
+ErrorCode BufferPool::open( const std::string& path ) noexcept {
+	ErrorCode error = p_storage->open(path);
+	if ( error != ErrorCode::E_NO_ERROR ) {
+		return error;
+	}
 
+	p_pool = (char*) malloc( m_config.m_poolSizeKB*1024*sizeof(char) );
 	uint32_t pageSizeKB = p_storage->getPageSize() / 1024;
-	uint32_t poolElems = config.m_poolSizeKB / pageSizeKB;
+	uint32_t poolElems = m_config.m_poolSizeKB / pageSizeKB;
 	m_descriptors.resize(poolElems);
-	m_allocationTable.resize(poolElems);
 	m_nextCSVictim = 0;
+
+	error = loadAllocationTable();
+	if ( error != ErrorCode::E_NO_ERROR ) {
+		return error;
+	}
+
+	return ErrorCode::E_NO_ERROR;
 }
 
-ErrorCode BufferPool::alloc( BufferHandler* bufferHandler ) noexcept {
-	// Get an empty buffer pool slot for the page.
-	bufferId_t bId;
-	ErrorCode error = getEmptySlot(&bId);
+ErrorCode BufferPool::create( const std::string& path, const FileStorageConfig& config, const bool& overwrite ) noexcept {
+	ErrorCode error = p_storage->create(path, config, overwrite);
 	if ( error != ErrorCode::E_NO_ERROR ) {
 		return error;
 	}
 
-	// Reserve space in disk and get the corresponding pageId_t.
+	p_pool = (char*) malloc( m_config.m_poolSizeKB*1024*sizeof(char) );
+	uint32_t pageSizeKB = p_storage->getPageSize() / 1024;
+	uint32_t poolElems = m_config.m_poolSizeKB / pageSizeKB;
+	m_descriptors.resize(poolElems);
+	m_nextCSVictim = 0;
+
+	return ErrorCode::E_NO_ERROR;
+}
+
+ErrorCode BufferPool::close() noexcept {
+	ErrorCode error = checkpoint();
+	if ( error != ErrorCode::E_NO_ERROR ) {
+		return error;
+	}
+
+	error = p_storage->close();
+	if ( error != ErrorCode::E_NO_ERROR ) {
+		return error;
+	}
+
+	free(p_pool);
+	m_descriptors.clear();
+	m_allocationTable.clear();
+	m_bufferToPageMap.clear();
+
+	return ErrorCode::E_NO_ERROR;
+}
+
+ErrorCode BufferPool::alloc( BufferHandler* bufferHandler ) noexcept {
+	bufferId_t bId;
 	pageId_t pId;
-	error = p_storage->reserve(1, &pId);
+
+	// Get an empty buffer pool slot for the page.
+	ErrorCode error = getEmptySlot(&bId);
 	if ( error != ErrorCode::E_NO_ERROR ) {
 		return error;
 	}
 
+	// Look if there is an unallocated disk page.
+	bool found = false;
+	for (uint64_t i = 0; i < p_storage->size() && !found; ++i) {
```

Comment:
This is very inefficient and should be fixed. As a database grows, since normally there are more allocations than releases, the cost of finding a free page will grow linearly as the size of the databases increases. One way of fixing this would be to have two collections: the current allocationTable and a list of freed pages. When looking for a free page, one is pulled from the freed pages collection if any, otherwise, the next free page from the end of the file is picked.

Reply:
Fixed in bd2886e.